File: MATLAB_Sim/Quat2Euler.m
function [Euler] = Quat2Euler(q)

q1 = q(1,:);
q2 = q(2,:);
q3 = q(3,:);
q4 = q(4,:);

phi = atan2(2*(q2.*q3+q1.*q4),1-2*(q1.^2+q2.^2));
temp = 2*(q1.*q3+q4.*q2);
gr1 = temp>1;
le1 = temp<-1;
temp(gr1)=1;
temp(le1)=-1;
theta = asin(temp);

% theta = atan2(-sin(phi).*(q4.*q2+q3.*q1),(q2.*q3-q1.*q4));

psi = atan2(2*(q1.*q2+q3.*q4),1-2*(q2.^2+q3.^2));

Euler = [phi;theta;psi];
end

File: MATLAB_Sim/ccodeFiles/Quat2Euler.c


float t2 = q2*q2;
float  t3 = t2*2.0;
float  t4 = -t3;
A0(0,0) = atan2(q1*q4*2.0+q2*q3*2.0,t4-(q1*q1)*2.0+1.0);
float temp = asin(q1*q3*2.0+q2*q4*2.0);
if (temp > 1.0){ temp = 1.0; }
if (temp < -1.0){ temp = -1.0; }
A0(1,0) = temp;
A0(2,0) = atan2(q1*q2*2.0+q3*q4*2.0,t4-(q3*q3)*2.0+1.0);
